### PR TITLE
Added support for getting the type of a GUI node at runtime

### LINKED ISSUE #9358: Add function to get gui node type
**Is your feature request related to a problem? Please describe (REQUIRED):**
I have a code to center label and icon.
For example icon and price.

Now i use hack to check if node text.
```lua
gui.get_text(node) ~= nil 
```
```lua
local M = {}

local TEXT_NODE_METRICS = {
    tracking = 1,
    line_break = false,
    width = 0
}

local function is_text_node(node)
    return gui.get_text(node) ~= nil
end

local function get_text_node_size(node)
    local font = gui.get_font(node)
    local text = gui.get_text(node)

    TEXT_NODE_METRICS.tracking = gui.get_tracking(node)
    TEXT_NODE_METRICS.line_break = gui.get_line_break(node)
    TEXT_NODE_METRICS.width = gui.get_size(node).x

    local metrics = resource.get_text_metrics(gui.get_font_resource(font), text, TEXT_NODE_METRICS)

    return metrics.width
end

function M.set_nodes_to_center(l_node, r_node, delta)
    delta = delta or 0
    
    local l_size_x = (is_text_node(l_node) and get_text_node_size(l_node) or gui.get_size(l_node).x) * gui.get_scale(l_node).x
    local r_size_x = (is_text_node(r_node) and get_text_node_size(r_node) or gui.get_size(r_node).x) * gui.get_scale(r_node).x
    local l_pivot = gui.get_pivot(l_node)
    local r_pivot = gui.get_pivot(r_node)
    local l_pos = gui.get_position(l_node)
    local r_pos = gui.get_position(r_node)

    local text_length = l_size_x + r_size_x + delta
    local l_dx = (text_length / 2) - l_size_x
    local r_dx = (text_length / 2) - r_size_x

    l_pos.x = -l_dx
    if l_pivot == gui.PIVOT_W or l_pivot == gui.PIVOT_NW or l_pivot == gui.PIVOT_SW then
        l_pos.x = l_pos.x - l_size_x
    elseif l_pivot == gui.PIVOT_CENTER then
        l_pos.x = l_pos.x - l_size_x / 2
    end
    r_pos.x = r_dx
    if r_pivot == gui.PIVOT_E or r_pivot == gui.PIVOT_NE or r_pivot == gui.PIVOT_SE then
        r_pos.x = r_pos.x + r_size_x
    elseif r_pivot == gui.PIVOT_CENTER then
        r_pos.x = r_pos.x + r_size_x / 2
    end

    gui.set_position(l_node, l_pos)
    gui.set_position(r_node, r_pos)
    return text_length, l_pos, r_pos, l_size_x, r_size_x
end

return M

```

**Describe the solution you'd like (REQUIRED):**
Add gui.get_node_type() or something like that

**Describe alternatives you've considered (REQUIRED):**
```lua
local function is_text_node(node)
    return gui.get_text(node) ~= nil
end
```

**Additional context (OPTIONAL):**

### PR DESCRIPTION
**Pull request by ghost**
Added support for getting the type of a GUI node at runtime.

The `gui.get_type()` function returns one of the following constants:

* `gui.TYPE_BOX`
* `gui.TYPE_TEXT`
* `gui.TYPE_PIE`
* `gui.TYPE_PARTICLEFX`
* `gui.TYPE_CUSTOM`

If the node uses a custom type, then the `hash` name of the custom type is returned as a secondary value.

```lua
local node = gui.new_box_node(vmath.vector3(100, 100, 0), vmath.vector3(20, 20, 1))
local type, subtype = gui.get_type(node)
pprint(type, subtype) -- 0 (gui.TYPE_BOX), nil
```

Resolves #9358 

## PR checklist

* [x] Code
	* [x] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [x] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [x] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

**Comment by VowSoftware** [comment]
Test is passing for box nodes, text nodes, and pie nodes. However, I don't know how to add particlefx nodes and custom nodes to the test. Could I get some guidance on that? Custom nodes are particularly important because of the `subtype` return value.

Two more general questions about the PR process.

> Make sure that manuals are updated (in github.com/defold/doc)

I believe I updated all of the relevant comments in the code itself for generating documentation. Is that different from updating the manuals at github.com/defold/doc?

> Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.

I think this is referring to the Kanban projects in the Defold repository. How do I know which project I should assign this issue to?

**Comment by JCash** [comment]
I think the code looks good an as expected.

For creating the pfx/custom node, I recommend doing it from C, and then using `gui.get_node(my_c_id)`.
I think it's especially more easy for the custom node, for which we don't have a "gui.new_" function for.

**Comment by JCash** [comment]
Also, the PR is difinitely ready to come out of draft mode :)

**Comment by JCash** [comment]
Regarding the documentation, that's part of the review process. We'll review the code to see if it contains the documentation and/or has any issues. MOve the PR from draft into review mode.

Regarding "Assign to project", you cannot as you're not an admin of the repository/project.

File: engine/gui/src/gui_script.cpp [hunk 1]
```diff
@@ -551,6 +551,48 @@ namespace dmGui
...
+     * @param node [type:node] node from which to get the type
+     * @return type [type:constant] type
+     *
+     * - `gui.TYPE_BOX`
+     * - `gui.TYPE_TEXT`
+     * - `gui.TYPE_PIE`
+     * - `gui.TYPE_PARTICLEFX`
+     * - `gui.TYPE_CUSTOM`
+     * 
+     * @return subtype [type:hash|nil] name of custom type
```
Comment (JCash) [inline engine/gui/src/gui_script.cpp:566]: Regarding "hash" here, I think it's a bit confusing for the user, as they then might expect our Lua hash (the 64 bit hash user type).

I think it's better to say "[type: integer|nil] Id of the custom type"

Comment (VowSoftware) [reply engine/gui/src/gui_script.cpp:566]: Okay, I think I understand. I'm not sure what the difference is yet between the different hash types yet. I referenced the custom node type from [this spine example](https://github.com/defold/extension-spine/blob/cab4430db95dec3fe3c89c7d5356f37f533d32eb/defold-spine/src/script_spine_gui.cpp#L25), which uses:

```c
static const uint32_t SPINE_NODE_TYPE = dmHashString32("Spine");
```

Is that not the same thing as the `hash: ["Spine"]` that shows up when you print a hash value in the editor?

**Review by JCash — APPROVED** [review]

**Pull request merged**